b3sum/src/main.rs

```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
     }
 }
 
+fn check_file(input: impl Read, mmap_disabled: bool) -> Result<u64> {
+    let hasher = blake3::Hasher::new();
+    let mut num_failed = 0;
+
+    let reader = std::io::BufReader::new(input);
+    for (line_num, line) in reader.lines().enumerate() {
+        // Read the line and trim it
+        let line: String = line.with_context(|| format!("line {}", line_num + 1))?;
+        let line = line.trim();
+
+        // Skip the line if it's a comment or empty
+        if line.starts_with('#') {
+            continue;
+        }
+        if line.len() == 0 {
+            continue;
+        }
+
+        // Split line by whitespace
+        let mut line = line.split_whitespace();
+
+        // Get the hash value and parse the hex value
+        let hash: &str = line.next().expect("Unexpected missing value");
+        let hash = hex::decode(hash)
+            .with_context(|| format!("error decoding hex on line {}", line_num + 1))?;
+
+        // Get the filename
+        let filename: &str = line
+            .next()
+            .ok_or_else(|| Error::msg(format!("missing filename on line {}", line_num + 1)))?;
+        let os_filename: std::ffi::OsString = filename.into();
+
+        // Hash the file
+        let hasher = hasher.clone();
+        let mut check_hash = Vec::<u8>::with_capacity(hash.len());
+        let check_reader = hash_file(&hasher, &os_filename, mmap_disabled)
+            .with_context(|| format!("filename: {}", filename))?;
+        check_reader
+            .take(hash.len() as u64)
+            .read_to_end(&mut check_hash)?;
```

Comment:
I'm a little skeptical of deriving the length from the input hashes. Say for example that the input hashes are just 1 byte long. That could lead to a situation where you think you're checking file integrity, but you're really not checking much. This gets more problematic if it's possible for some attacker to _control_ the length of the hashes you check.

On the flip side, checking hashes longer than 32 bytes doesn't add any security, but people might get confused and think that it does.

Maybe just have `--check` always assume the default length? Maybe even conflict with `--length`? (Incidentally it should probably also conflict with `--keyed` and `--derive-key`.)

Reply:
I was trying to ensure that all output from `b3sum` is also valid input for `b3sum --check`.    So that if you do `b3sum * --length=128` and pass that back into `b3sum --check` it should work. 

I would recommend two things:

1. Throw a warning if the hash lengths are under some threshold at which we consider the security compromised.  We should do this for both checking and generating the checksum file. 

2. Add a "strict" mode where warnings are upgraded to errors. 

This has some precedence.  `b2sum-rs` has the following options that we could emulate:
```
The following five options are useful only when verifying checksums:
      --ignore-missing  don't fail or report status for missing files
      --quiet           don't print OK for each successfully verified file
      --status         don't output anything, status code shows success
      --strict          exit non-zero for improperly formatted checksum lines
``` 

However, if we do decide to error on checking non-32-byte checksums, then we should make sure we at least generate a warning when generating them.

Reply:
I think if we have `check` throw an error saying "this only works with 32-byte hashes", we'll get 90% of the value with 10% of the effort. Trying to determine "security compromised" in general is too vague, and it depends on the application. (Collision-resistant checksums need to be 256 bits, but MACs can usually get away with 128, for example.)